File: List/List.java

public class List<T> {
    // Number of items in list
    private int size;

    // Capcity of the list
    private int capacity;

    // Array that holds the list items
    private T[] items;

    public List() {
        this.size = 0;
        this.capacity = 1;
        this.items = (T[]) new Object[capacity];
    };

    public void add(T v) {
        // Increase size and check that new size fits capacity
        if (++this.size > this.capacity) {
            this.capacity *= 2;
        }

        // If capacity changed resize array
        if (this.capacity > this.items.length) {
            T[] tmp = (T[]) new Object[this.capacity];
            for (int i = 0; i < this.size - 1; i++) {
                tmp[i] = this.items[i];
            }
                
            this.items = tmp;
        }
        // Assign new element
        this.items[this.size - 1] = v;
    };

    public void display() {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < this.size; i++) {
            sb.append(items[i]).append(" ");
        }
            
        System.out.println(sb.toString());
    };

    public int size() {
        return this.size;
    };

    public T get(int idx) {
        return this.checkIndexBounds(idx) 
            ? this.items[idx] 
            : null;
    };

    public int indexOf(T t) {
        for (int i = 0; i < this.size; i++) {
            if (t == items[i]) {
                return i;
            }
        }   
        
        return -1;
    };

    public void remove(int idx) {
        if (this.checkIndexBounds(idx)) {
            for (int i = idx; i < this.size; i++) {
                items[i] = items[i + 1];
            }
                
            items[--this.size] = items[this.capacity - 1];
        }
    };

    public void remove(T t) {
        int idx = this.indexOf(t);
        if (idx != -1) {
            this.remove(idx);
        }
    };

    public boolean isEmpty() {
        return this.size == 0;
    };

    public boolean contains(T t) {
        return this.indexOf(t) != -1;
    };

    public T first() {
        return this.isEmpty() ? null : items[0];
    };

    public T last() {
        return this.isEmpty() ? null : items[this.size - 1];
    };

    public void set(int idx, T t) {
        if (this.checkIndexBounds(idx)) {
            items[idx] = t;
        }
    };

    public int capacity() {
        return this.capacity;
    };

    private boolean checkIndexBounds(int idx) {
        return idx >= 0 && idx <= this.size;
    }
};

File: List/List.cs

namespace list
{
    public class List<T>
    {
        // Number of items in list
        private int len;

        // Capacity of the list
        private int capacity;

        // Array that holds the list items
        private T[] items;

        public List() {
            len = 0;
            capacity = 1;
            items = new T[capacity];
        }

        public void Add(T v) {
            // Increase size and check that new size fits capacity
            if (++len > capacity) capacity *= 2;
            
            // If capacity changed; resize array
            if (capacity > items.Length)
            {
                T[] tmp = new T[capacity];
                for (int i = 0; i < len - 1; i++)
                    tmp[i] = items[i];
                items = tmp;
            }

            // Assign new element
            items[len - 1] = v;
        }   

        public void Display() {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            for (int i = 0; i < len; i++)
                sb.Append(items[i]).Append(" ");
            System.Console.WriteLine(sb.ToString());
        }

        public int Size() {
            return len;
        }

        public int Capacity() {
            return capacity;
        }

        public T Get(int idx) {
            return idx <= len ? items[idx] : default(T);
        }

        public int IndexOf(T t) {
            for (int i = 0; i < len; i++)
                if (t.Equals(items[i]))
                    return i;
            
            return -1;
        }
    
        public void Remove(int idx) {
            if (0 <= idx && idx <= len)
            {
                for (int i = 0;  i < len; i++)
                    items[i] = items[i + 1];

                items[--len] = items[capacity - 1];
            }
        }

        public void Remove(T t) {
            int idx = IndexOf(t);
            if (idx != -1) Remove(idx);
        }

        public bool IsEmpty() {
            return len == 0;
        }

        public bool Contains(T t) {
            return IndexOf(t) != -1;
        }

        public T First() {
            return IsEmpty() ? default(T) : items[0];
        }

        public T Last() {
            return IsEmpty() ? default(T) : items[len - 1];
        }

        public void Set(int idx, T t) {
            if (0 <= idx && idx < len)
                items[idx] = t;
        }
    }
}